### ISSUE TITLE
wtclient attempts to connect to tower over tor despite being on 127.0.0.1

### ISSUE THREAD

**Issue by 3nprob** [lightningnetwork/lnd#5647]
### Background

LND node with Tor enabled and watchtower enabled. Trying to add the local tower results in trying to connect over tor, which obviously fails. I see hree ways to address this:

1. Easy trivial: Don't route connections to local subnet over tor
2. Global: Add configuration field `--tor.ignore=`, listing subnets to not route over tor
3. Specific: Add configuration field `--wtclient.tor_ignore=`, listing subnets to not route over tor for wtclient

The node connects to bitcoind on localhost, so this seems to be wtclient-specific

```
$ lncli tower info
{
    "pubkey": "deadbeef",
    "listeners": [
        "127.0.0.1:9911"
    ],
    "uris": [
        "deadbeef@127.0.0.1:9911"
    ]
}


$ lncli wtclient add deadbeef 

{

}

$ lncli wtclient towers
{
    "towers": [
        {
            "pubkey": "deadbeef",
            "addresses": [
                "127.0.0.1:9911"
            ],
            "active_session_candidate": true,
            "num_sessions": 0,
            "sessions": [
            ]
        }
    ]
}
```

Log:
```
2021-08-21 04:27:16.550 [DBG] WTCL: (anchor) Attempting session negotiation with tower=deadbeef
2021-08-21 04:27:16.550 [DBG] WTCL: (legacy) Attempting session negotiation with tower=deadbeef
2021-08-21 04:27:16.578 [DBG] WTCL: (anchor) Request for session negotiation with tower=deadbeef@127.0.0.1:9911 failed, trying again -- reason: socks connect tcp 127.0.0.1:9050->127.0.0.1:9911: unknown error general SOCKS server failure
2021-08-21 04:27:16.578 [DBG] WTCL: (anchor) Session negotiation with tower=deadbeef failed, trying again -- reason: session negotiation unsuccessful
2021-08-21 04:27:16.580 [DBG] WTCL: (legacy) Request for session negotiation with tower=deadbeef@127.0.0.1:9911 failed, trying again -- reason: socks connect tcp 127.0.0.1:9050->127.0.0.1:9911: unknown error general SOCKS server failure
2021-08-21 04:27:16.580 [DBG] WTCL: (legacy) Session negotiation with tower=deadbeef failed, trying again -- reason: session negotiation unsuccessful
```

### Your environment

* version of `lnd`: v0.13.1-beta

**Comment by 3nprob** [comment]
In an attempt to connect over the HS instead, it doesn't seem like the WT HS is initialized properly. Removing the following from config:
```
watchtower.externalip=127.0.0.1
```

Still after restart:

```
$ lncli tower info
{
    "pubkey": "deadbeef",
    "listeners": [
        "127.0.0.1:9911"
    ],
    "uris": [
    ]
}
```

**Comment by 3nprob** [comment]
Oddly enough, i see no logs at all from `WTWR` itself despite setting the level to `trace`:

`[INF] LTND: Version: 0.13.1-beta commit=v0.13.1-beta, build=development, logging=default, debuglevel=debug,PEER=info,SRVR=info,DISC=info,WTWR=trace`

The configuration settings for it  obviously do take effect, though

**Comment by Roasbeef** [comment]
I think this'll be addressed by https://github.com/lightningnetwork/lnd/pull/5410?

**Comment by 3nprob** [comment]
God damn it I swear I was looking for issues before submitting - but yes, it should.

**Comment by 3nprob** [comment]
@Roasbeef Almost, but not quite. I just want connections to `127.0.0.1` to skip tor

**Comment by Roasbeef** [comment]
> The node connects to bitcoind on localhost, so this seems to be wtclient-specific

If this is the case, then you likely don't need to mess with the main Tor behavior at all, and can instead just examine the two connection sites to see why one uses the Tor path and the other doesn't.

**Comment by Roasbeef** [comment]
Check out this documentation for reaching the tower over a hidden service: https://github.com/lightningnetwork/lnd/blob/master/docs/watchtower.md#tor-hidden-services

You can drop most of the config options you've specified above.

**Comment by 3nprob** [comment]
> Check out this documentation for reaching the tower over a hidden service: https://github.com/lightningnetwork/lnd/blob/master/docs/watchtower.md#tor-hidden-services
> 
> You can drop most of the config options you've specified above.

For sure. I did have this working; the idea here is to not expose it at all in interest of preserving resources and minimize exposure area.

As you note my PR goes way beyond the specific issue here (which was intended as a bug report as opposed to a feature request) but while I was mucking around the code I thought I might as well fill in some missing functionality that I need myself and assume will be useful for others :)

**Comment by Roasbeef** [comment]
Ah actually, re-reading this again, this just seems to be an general issue in that we should always skip `localhost` for outbound connections when Tor is active. IIRC, we resolved this issue when trying to to _DNS resolutions_ over Tor, but I guess I thought this was addressed there as well.

**Comment by Roasbeef** [comment]
Appears to be related to: https://github.com/lightningnetwork/lnd/issues/4191

**Comment by 3nprob** [comment]
@Roasbeef Yes, but also in general it can be desirable to do the same for local subnets (e.g. for nodes or services on the local LAN, when using software-defined networking, etc)

I think #5651 should cover any scenario with unsurprising defaults (skip tor only for `localhost`, `127.*.*.*`. `::1`).

**Comment by openoms** [comment]
Just making a note here about the same issue when trying to connect to a neutrino node over the localhost or LAN.

the `tor.skip-proxy-for-clearnet-targets=true` will be very useful for the not privacy oriented nodes which just want access to the Tor network as well, but it poses the problem of revealing the IP addresses for regular home users.

Being able to specify services (like the `wtclient` or `neutrino`) to skip the Tor proxy could be a solution.

**Comment by 3nprob** [comment]
@openoms If you patch with #5651 you could just add `tor.no-proxy-target=$NEUTRINO_IP/32`.

I don't really see a scenario where you would actually want to skip tor for a given IP only for certain services/modules but not for others, and I think users who want to selectively skip tor will know which IPs/subnets are intended statically.

Or do you have some case you feel is not covered?

**Comment by openoms** [comment]
thanks @3nprob that covers my usecase. Agree that the tor/clearnet connection should be selective to IP addresses rather than services. Great PR!

**Comment by maximlomans** [comment]
Thanks 3nprob fixed for me too ..
